### PR TITLE
Revert blockchain in shard when epoch cross check is wrong

### PR DESCRIPTION
**Pull request by sasurobert**
fixing situation when metachain does not accept blocks from shard because of wrong epoch

### CODE REVIEW

File: process/block/shardblock.go [hunk 1]
```diff
@@ -879,6 +880,44 @@ func (sp *shardProcessor) CommitBlock(
...
+		return nil
+	}
+
+	shouldRevertChain := false
+	nonce := currentHeader.GetNonce()
+	shouldEnterNewEpochRound := sp.epochStartTrigger.EpochFinalityAttestingRound() + process.EpochChangeGracePeriod
+
+	for round := currentHeader.GetRound(); round > shouldEnterNewEpochRound && currentHeader.GetEpoch() != sp.epochStartTrigger.Epoch(); round = currentHeader.GetRound() {
+		shouldRevertChain = true
+		nextHeader, _, err := process.GetHeaderFromStorageWithNonce(
```
Comment (SebastianMarian) [inline process/block/shardblock.go]: prevHeader

File: process/sync/baseForkDetector.go [hunk 2]
```diff
@@ -32,6 +32,7 @@ type forkInfo struct {
 	lastProposedBlockNonce uint64
 	shouldForceFork        bool
 	isNotarizedShardStuck  bool
+	forcedForkNonce        uint64
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/sync/baseForkDetector.go]: done

File: process/sync/baseForkDetector.go [hunk 3]
```diff
@@ -46,6 +47,20 @@ type baseForkDetector struct {
 	blackListHandler process.BlackListHandler
 }
 
+func (bfd *baseForkDetector) SetForkNonce(nonce uint64) {
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/sync/baseForkDetector.go]: done

File: process/sync/baseSync.go [hunk 4]
```diff
@@ -591,6 +591,11 @@ func (boot *baseBootstrap) rollBack(revertUsingForkNonce bool) error {
 		if !revertUsingForkNonce && currHeader.GetNonce() <= boot.forkDetector.GetHighestFinalBlockNonce() {
 			return ErrRollBackBehindFinalHeader
 		}
+		shouldEndRollBack := revertUsingForkNonce && currHeader.GetNonce() < boot.forkInfo.Nonce
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/sync/baseSync.go]: done

File: process/block/shardblock.go (hunk 1 above)
Comment (sasurobert) [reply process/block/shardblock.go]: done

File: process/sync/shardForkDetector.go [hunk 5]
```diff
@@ -38,7 +38,7 @@ func NewShardForkDetector(
 	checkpoint := &checkpointInfo{}
 	bfd.setFinalCheckpoint(checkpoint)
 	bfd.addCheckpoint(checkpoint)
-	bfd.fork.forcedForkNonce = math.MaxUint64
+	bfd.fork.nonce = math.MaxUint64
```
Comment (SebastianMarian) [inline process/sync/shardForkDetector.go]: You forgot to rename this in metaForkDetector

Comment (sasurobert) [reply process/sync/shardForkDetector.go]: done

**Pull request merged**